storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3109,10 +3109,1760 @@ buf_block_t *fsp_get_latched_xdes_page(
+    if (rec && rec != page + PAGE_OLD_SUPREMUM)
+    {
+      ulint len;
+      ulint offset=
+        rec_get_nth_field_offs_old(rec, rec_get_n_fields_old(rec) - 1,
+                                   &len);
+      ut_ad(len == 4);
+
+      ut_ad(offset < srv_page_size);
+      child_page_no= mach_read_from_4(rec + offset);
+    }
+    else
+    {
+      child_page_no= FIL_NULL;
+      err= DB_CORRUPTION;
+      break;
+    }
+    if (cur_level == level + 1)
+      break;
+  }
+  mtr.commit();
+  return child_page_no;
+}
+
+dberr_t indexDefragmenter::defragment_level(uint16_t level,
+                                            spaceDefragmenter *space_defrag) noexcept
+{
+  dberr_t err= DB_SUCCESS;
+  uint32_t cur_page_no= get_level_block(level, &err);
+  if (cur_page_no == FIL_NULL)
+    return err;
+
+  fil_space_t *space= fil_system.sys_space;
+  uint32_t extent_size= space_defrag->m_extent_size;
+  mtr_t mtr;
+  mtr.start();
+  mtr.x_lock_space(space);
+  mtr_x_lock_index(m_index, &mtr);
+
+  buf_block_t *block= buf_page_get_gen(page_id_t{0, cur_page_no},
+                                       0, RW_X_LATCH, nullptr,
+                                       BUF_GET_POSSIBLY_FREED, &mtr, &err);
+  if (!block)
+  {
+err_exit:
+    mtr.commit();
+    return err;
+  }
+
+  while (block)
+  {
+    page_t *page= buf_block_get_frame(block);
+    uint32_t next_page_no= btr_page_get_next(page);
+    uint32_t cur_extent= (cur_page_no / extent_size) * extent_size;
+    uint32_t old_state= space_defrag->get_state(cur_extent);
+
+    if (old_state == XDES_FREE)
+    {
+fetch_next_page:
+      mtr.commit();
+
+      if (next_page_no == FIL_NULL)
+        break;
+      cur_page_no= next_page_no;
+
+      mtr.start();
+      mtr_x_lock_index(m_index, &mtr);
+      mtr.x_lock_space(fil_system.sys_space);
+      block= fsp_get_latched_page(page_id_t{0, cur_page_no},
+                                  &mtr, &err);
+      continue;
+    }
+
+    uint32_t new_extent= space_defrag->get_new_extent(cur_extent);
+    /* There is no need for extent to be changed */
+    if (new_extent == cur_extent)
+    {
+      if (level)
+      {
+        err= store_child_pages(block);
+	if (err) goto err_exit;
+      }
+      goto fetch_next_page;
+    }
+
+    buf_block_t *header_block=
+      fsp_get_latched_page(page_id_t{0, 0}, &mtr, &err);
+    if (!header_block)
+      goto err_exit;
+
+    buf_block_t *root= btr_root_block_get(m_index, RW_SX_LATCH, &mtr,
+                                          &err);
+    if (UNIV_UNLIKELY(!root))
+      goto err_exit;
+
+    fseg_header_t *seg_header=
+      root->page.frame + (level
+                          ? PAGE_HEADER + PAGE_BTR_SEG_TOP
+                          : PAGE_HEADER + PAGE_BTR_SEG_LEAF);
+
+    buf_block_t *iblock;
+    fseg_inode_t *inode= fseg_inode_try_get(seg_header, 0, 0, &mtr,
+                                            &iblock, &err);
```

Comment:
Do we have to acquire the root block inside the loop? Actually, `seg_header` could be `const`. If we were continuously holding an exclusive index latch (instead of repeatedly acquiring and releasing across `mtr` restarts), then no concurrent thread should be allowed to modify the root block. Could we acquire the root block outside `mtr` with `buf_pool.page_fix()` (and remember to `root->unfix()` at the end) if it is read-only during the entire operation?